### ISSUE TITLE
Better querying for existing NWB's 

### ISSUE THREAD

**Issue by rachelstephlee** [AllenNeuralDynamics/aind-data-access-api#109]
# **Is your feature request related to a problem? Please describe.**

I want to be able to search for fiber photometry sessions in S3 and Code ocean easily. I care about this for fiber photometry for the dynamic foraging task, but this is important across tasks and modalities. 

This is important. Soon we will have a data management problem that if we don't get a head start on, we will be drowning in data but no way to search for and sort through it. It will be more frustrating when we have more data. Best to start now and have some mini infrastructure going. 

# **Describe the solution you'd like**

A programmatic way to search for sessions I want, or a GUI to do so. The former is much preferred over the latter. I have code below that searches for the relevant fields including: 
- *subject ID
- session date 
- *task type
- stage of task (as in, what stage the mouse is trained in)
- *if it has been processed by a pipeline, and what version of the pipeline
- *surgery/probe/virus information (see https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/252)
- *data asset ID for unprocessed data, data asset ID for processed NWB

some of these fields might not be searchable via docDB (they're taken from han's code that tracks data separately)

I get that this is less of a feature request and more like a whole story to address that requires non-trivial sci-comp support. But I'd like to point out that-- the sooner we start writing some functions for this when we only have 5000 sessions (rather than 10K sessions), the cheaper and more non-trivial this feature would be. I've starred the most important fields in my list above. 

## **nice to have:** 
- names of the unprocessed data, names of the processed NWB
- any data description/notes for the NWB. 
- foraging efficiency, which is a measure han used to look at mouse performance. 

## **later:** 
- histology information. This is already stored some place but it's not organized. 
- a way to easily track mouse performance (i.e. a better measure than foraging efficiency)

# **Describe alternatives you've considered**

We have streamlit right now. It is not an ideal solution and querying should be implemented by sci-comp, not maintained by Han. 
- also streamlit slows down each day as we add ever more sessions that it slows and lags. 

# **Additional context**

I have code here that queries DocDB for sessions. 

https://github.com/AllenNeuralDynamics/foraging-behavior-browser/blob/main/code/util/fetch_data_docDB.py

Han also has written a whole doc describing what we're looking for eventually.

**Comment by hanhou** [comment]
I think we should first make sure all new and old data in CO have the correct metadata (see https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/373). If not, we should re-process or backfill. Once this is done, querying nwb from docDB should be much easier.

**Comment by hanhou** [comment]
In the meantime, I'd like to request a tool to help us navigate the NWBs like this:

![IMG_2629](https://github.com/user-attachments/assets/f6258f42-1e4d-476a-9462-d57fd8480d40)

Briefly, it allows users to try and compare arbitrary queries in docDB. Here is a breakdown of the panels:

- `Field hints`: 
    - drop down menu shows which fields are available for query in our metadata
- `Query`: 
    - users can load presets (part of what Rachel has requested) or try their own arbitrary queries, and the preview will be shown in `Query Preview`
    - users can assign an alias to a query and add it to the `Query Comparator`.
- `Query Comparator`:
    - show Venn diagram or [Upset Plot](https://en.wikipedia.org/wiki/UpSet_plot) of selected queries, which is a great way to overview our data
- `Query Preview`:
    - show preview of the queried results from `Query` or when the user selects any segment of the Venn diagram.
    - users can control which fields to project

I think this could be a very useful addition to [Dan's metadata portal](https://metadata-portal.allenneuraldynamics.org/app).  @dyf  @dbirman @saskiad  @rachelstephlee

**Comment by dbirman** [comment]
I'm planning to add some limited query functionality to the portal https://github.com/AllenNeuralDynamics/aind-metadata-viz/issues/30, the intention is that it will also provide you with the mongodb query code that you can then use with the data access API directly. It will probably be simpler than what you've outlined Han at least to start.

More generally @rachelstephlee I think what you're requesting here already exists through the data access api, but maybe not in an ideal format for your purpose? It's already possible write a query that combines the various fields you described and you'll get back the records you want. For example, a query like https://api.allenneuraldynamics.org/v1/metadata_index/data_assets?filter={%22subject.subject_id%22:%22711042%22}&limit=1 could be chained with a window of possible dates, etc. Are you requesting that we make a more abstract API that allows you to do this without knowing how to write the mongodb query itself?

**Comment by rachelstephlee** [comment]
thanks for all the helpful discussion!

agreed with han that this should wait until all metadata correction/updates come thru. 

dan, it probably would be much easier if you, me, and han sit down and sketch out what exactly we want (the minimum requirements, how expensive it would be) rather than discuss it via github here.

**Comment by rachelstephlee** [comment]
Per discussion with David, I mostly need some combination of better documentation and/or more user-friendly access methods. 

Like Han said, we need to finish fixing the metadata problems though (in particular, we don't have reliable metadata on the recording type) https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/373


But more discussion needs to be done for 1. at what point are we satisfied with reliable metadata for the past vs the future? We cannot wait for metadata to be perfect to have search usable 2. what amount of hand-holding do we need from sci-comp to have easily accessible docDB. @hanhou knows how to search using docDB, but I don't think that's true for most Scientists at AIND !

**Comment by hanhou** [comment]
>1. at what point are we satisfied with reliable metadata for the past vs the future? We cannot wait for metadata to be perfect to have search usable 

I strongly agree. A more accessible tool also allows us to do test-driven development/fixing of our metadata. I quickly prototyped something similar to the app I proposed above in Streamlit. For example, in this venn diagram below, the red circle should be inside the green circle and overlap perfectly with the blue circle (see https://github.com/AllenNeuralDynamics/aind-physio-arch/issues/373).

<img width="933" alt="image" src="https://github.com/user-attachments/assets/4740371e-0489-4c24-a568-a078e01795ed" />

Here are [the queries](https://github.com/AllenNeuralDynamics/foraging-behavior-browser/blob/0e7e21cd31f702971444bd6a99f9d8e030c378cd/code/pages/3_AIND%20data%20access%20playground.py#L39-L60) I used to generate the venn diagram. We can use such a tool to sanity check and fix our metadata for now, and once the metadata becomes "perfect", we can use it to catalogue our sessions and provide a live documentation for new metadata users.

**Comment by hanhou** [comment]
> @hanhou knows how to search using docDB, but I don't think that's true for most Scientists at AIND !

Sci-comp already has a good documentation for accessing docDB https://aind-data-access-api.readthedocs.io/en/latest/UserGuide.html#document-database-docdb, but we can only trust the queried results if we have perfect metadata.

**Comment by rachelstephlee** [comment]
I think we can either close this comment for now or we can put this on hold. Thanks for the useful discussion. 


For complete information, I have found out that [sci-comp has provided tutorials for docDB here also: ](https://codeocean.allenneuraldynamics.org/collections/d69a9a08-a6e0-4e86-88c5-b9edba3b9d9c?page=1&filter=all)
